clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java

```diff
@@ -281,27 +284,35 @@ private long sendProducerData(long now) {
     }
 
     private boolean maybeSendTransactionalRequest(long now) {
-        if (transactionManager != null && transactionManager.hasInflightRequest())
+        if (transactionManager == null || !transactionManager.isTransactional())
+            return false;
+
+        if (transactionManager.hasInflightRequest()) {
+            log.trace("There is already an inflight transactional request. Going to wait for the response.");
             return true;
+        }
 
-        if (transactionManager == null || !transactionManager.hasPendingTransactionalRequests())
+        if (!transactionManager.hasPendingTransactionalRequests()) {
+            log.trace("There are no pending transactional requests to send");
             return false;
+        }
 
         TransactionManager.TxnRequestHandler nextRequestHandler = transactionManager.nextRequestHandler();
 
-        if (nextRequestHandler.isEndTxn()) {
-            if (transactionManager.isCompletingTransaction() && accumulator.hasUnflushedBatches()) {
-                if (!accumulator.flushInProgress())
-                    accumulator.beginFlush();
-                transactionManager.reenqueue(nextRequestHandler);
-                return false;
-            } else if (transactionManager.isInErrorState()) {
-                nextRequestHandler.fatal(new KafkaException("Cannot commit transaction when there are " +
-                        "request errors. Please check your logs for the details of the errors encountered."));
-                return false;
-            }
+        if (nextRequestHandler.isEndTxn() && transactionManager.isCompletingTransaction() && accumulator.hasUnflushedBatches()) {
+            if (!accumulator.flushInProgress())
+                accumulator.beginFlush();
+            transactionManager.reenqueue(nextRequestHandler);
+            log.trace("Going to wait for pending ProducerBatches to flush before sending an end transaction request");
```

Comment:
nit: we should get in the habit of including the transactionalId in the messages which relate to transactions. That will make debugging easier when we have multiple producers in the same application.